pkg/daemon/pinned_image_set.go

```diff
@@ -0,0 +1,1278 @@
+		minStorageAvailableBytes: minStorageAvailableBytes,
+		featureGatesAccessor:     featureGatesAccessor,
+		queue:                    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "pinned-image-set-manager"),
+		prefetchCh:               make(chan prefetch, defaultPrefetchWorkers*2),
+		criClient:                criClient,
+		backoff: wait.Backoff{
+			Steps:    maxRetries,
+			Duration: retryDuration,
+			Factor:   retryFactor,
+			Cap:      retryCap,
+		},
+	}
+
+	imageSetInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    p.addPinnedImageSet,
+		UpdateFunc: p.updatePinnedImageSet,
+		DeleteFunc: p.deletePinnedImageSet,
+	})
+
+	mcpInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    p.addMachineConfigPool,
+		UpdateFunc: p.updateMachineConfigPool,
+		DeleteFunc: p.deleteMachineConfigPool,
+	})
+
+	p.syncHandler = p.sync
+	p.enqueueMachineConfigPool = p.enqueue
+
+	p.imageSetLister = imageSetInformer.Lister()
+	p.imageSetSynced = imageSetInformer.Informer().HasSynced
+
+	p.nodeLister = nodeInformer.Lister()
+	p.nodeListerSynced = nodeInformer.Informer().HasSynced
+
+	p.mcpLister = mcpInformer.Lister()
+	p.mcpSynced = mcpInformer.Informer().HasSynced
+
+	p.cache = lru.New(256)
+
+	return p
+}
+
+type imageInfo struct {
+	Name   string
+	Size   int64
+	Pulled bool
+}
+
+func (p *PinnedImageSetManager) sync(key string) error {
+	klog.V(4).Infof("Syncing PinnedImageSet for pool: %s", key)
+	node, err := p.nodeLister.Get(p.nodeName)
+	if err != nil {
+		return fmt.Errorf("failed to get node %q: %w", p.nodeName, err)
+	}
+
+	pools, _, err := helpers.GetPoolsForNode(p.mcpLister, node)
+	if err != nil {
+		return err
+	}
+
+	if len(pools) == 0 {
+		return nil
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), p.prefetchTimeout)
+	// cancel any currently running tasks in the worker pool
+	p.resetWorkload(cancel)
+
+	if err := p.syncMachineConfigPools(ctx, pools); err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			ctxErr := fmt.Errorf("requeue: prefetching images incomplete after: %v", p.prefetchTimeout)
+			p.updateStatusError(pools, ctxErr)
+			klog.Error(ctxErr)
+			return ctxErr
+		}
+		p.updateStatusError(pools, err)
+		return err
+	}
+
+	p.updateStatusProgressingComplete(pools, "All pinned image sets complete")
+	return nil
+}
+
+func (p *PinnedImageSetManager) syncMachineConfigPools(ctx context.Context, pools []*mcfgv1.MachineConfigPool) error {
+	images := make([]mcfgv1alpha1.PinnedImageRef, 0, 100)
+	for _, pool := range pools {
+		validTarget, err := validateTargetPoolForNode(p.nodeName, p.nodeLister, pool)
+		if err != nil {
+			klog.Errorf("failed to validate pool %q: %v", pool.Name, err)
+			return err
+		}
+		if !validTarget {
+			continue
+		}
+
+		if err := p.syncMachineConfigPool(ctx, pool); err != nil {
+			return err
+		}
+
+		// collect all unique images from all pools
```

Comment:
Did we mean to say something like `collect all unique images from PinnedImageSets targeted to the pool`

Reply:
nvm, I see that we are also looping over all pools.